src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen.py

```diff
@@ -0,0 +1,826 @@
+        # corresponds to doing no classifier free guidance.
+        do_classifier_free_guidance = guidance_scale > 1.0
+
+        # 3. Encode input prompt
+        prompt_embeds = self._encode_prompt(
+            prompt,
+            device,
+            num_images_per_prompt,
+            do_classifier_free_guidance,
+            negative_prompt,
+            prompt_embeds=prompt_embeds,
+            negative_prompt_embeds=negative_prompt_embeds,
+        )
+
+        # 4. Prepare timesteps
+        self.scheduler.set_timesteps(num_inference_steps, device=device)
+        timesteps = self.scheduler.timesteps
+
+        # 5. Prepare latent variables
+        num_channels_latents = self.unet.in_channels
+        latents = self.prepare_latents(
+            batch_size * num_images_per_prompt,
+            num_channels_latents,
+            height,
+            width,
+            prompt_embeds.dtype,
+            device,
+            generator,
+            latents,
+        )
+
+        def draw_inpaint_mask_from_boxes(boxes, size):
+            inpaint_mask = torch.ones(size[0], size[1])
+            for box in boxes:
+                x0, x1 = box[0] * size[0], box[2] * size[0]
+                y0, y1 = box[1] * size[1], box[3] * size[1]
+                inpaint_mask[int(y0) : int(y1), int(x0) : int(x1)] = 0
+            return inpaint_mask
+
+        # 5.1 Prepare GLIGEN variables
+        if gligen_phrases is not None:
+            assert len(gligen_phrases) == len(gligen_boxes)
+            assert batch_size == 1
+            max_objs = 30
+            _boxes = gligen_boxes
+            tokenizer_inputs = self.tokenizer(gligen_phrases, padding=True, return_tensors="pt").to(
+                self.text_encoder.device
+            )
+            _text_embeddings = self.text_encoder(**tokenizer_inputs).pooler_output
+            n_objs = min(len(_boxes), max_objs)
+            device = self.text_encoder.device
+            dtype = self.text_encoder.dtype
+            boxes = torch.zeros(max_objs, 4, device=device, dtype=dtype)
+            boxes[:n_objs] = torch.tensor(_boxes[:n_objs])
+            text_embeddings = torch.zeros(max_objs, 768, device=device, dtype=dtype)
+            text_embeddings[:n_objs] = _text_embeddings[:n_objs]
+            masks = torch.zeros(max_objs, device=device, dtype=dtype)
+            masks[:n_objs] = 1
+
+            repeat_batch = batch_size * num_images_per_prompt
+            if do_classifier_free_guidance:
+                repeat_batch = repeat_batch * 2
+            boxes = boxes.unsqueeze(0).expand(repeat_batch, -1, -1).clone()
+            text_embeddings = text_embeddings.unsqueeze(0).expand(repeat_batch, -1, -1).clone()
+            masks = masks.unsqueeze(0).expand(repeat_batch, -1).clone()
+            if do_classifier_free_guidance:
+                masks[: repeat_batch // 2] = 0
+            if cross_attention_kwargs is None:
+                cross_attention_kwargs = {}
```

Comment:
These blocks could benefit a lot from extensive comments. 

These blocks read quite dense. So, to supplement them, let's add descriptive comments to make sure they're as readable as possible. 

WDYT?